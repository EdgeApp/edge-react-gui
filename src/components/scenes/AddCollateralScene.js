// @flow

import s from '../../locales/strings.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { ManageCollateralScene } from './ManageCollateralScene.js'

type Props = {
  navigation: NavigationProp<'addCollateralScene'>,
  route: RouteProp<'addCollateralScene'>
}

export const AddCollateralScene = (props: Props) => {
  const { navigation, route } = props
  const { borrowEngine } = route.params

  return ManageCollateralScene({
    borrowEngine,
    defaultTokenId: borrowEngine.collaterals[0].tokenId,
    action: async req => await borrowEngine.deposit(req),
    actionWallet: 'fromWallet',
    ltvType: 'collaterals',
    ltvChange: 'decrease',

    showExchangeRateTile: true,
    showTotalDebtTile: true,
    showTotalCollateralTile: true,

    headerText: s.strings.loan_add_collateral,
    goBack: () => navigation.pop()
  })
}
