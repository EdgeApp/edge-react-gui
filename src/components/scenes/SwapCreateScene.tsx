import { gt, gte } from 'biggystring'
import { EdgeCurrencyWallet, EdgeSwapRequest, EdgeTokenId } from 'edge-core-js'
import * as React from 'react'
import { useState } from 'react'
import { Text, View } from 'react-native'
import { sprintf } from 'sprintf-js'

import { DisableAsset } from '../../actions/ExchangeInfoActions'
import { updateMostRecentWalletsSelected } from '../../actions/WalletActions'
import { getSpecialCurrencyInfo } from '../../constants/WalletAndCurrencyConstants'
import { useSwapRequestOptions } from '../../hooks/swap/useSwapRequestOptions'
import { useHandler } from '../../hooks/useHandler'
import { useWatch } from '../../hooks/useWatch'
import { lstrings } from '../../locales/strings'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { EdgeSceneProps } from '../../types/routerTypes'
import { getCurrencyCode } from '../../util/CurrencyInfoHelpers'
import { getWalletName } from '../../util/CurrencyWalletHelpers'
import { zeroString } from '../../util/utils'
import { EdgeAnim, fadeInDown30, fadeInDown60, fadeInDown90, fadeInUp60 } from '../common/EdgeAnim'
import { EdgeTouchableOpacity } from '../common/EdgeTouchableOpacity'
import { SceneWrapper } from '../common/SceneWrapper'
import { styled } from '../hoc/styled'
import { SwapVerticalIcon } from '../icons/ThemedIcons'
import { WalletListModal, WalletListResult } from '../modals/WalletListModal'
import { Airship, showError, showToast, showWarning } from '../services/AirshipInstance'
import { useTheme } from '../services/ThemeContext'
import { ExchangedFlipInputAmounts, ExchangedFlipInputRef } from '../themed/ExchangedFlipInput2'
import { LineTextDivider } from '../themed/LineTextDivider'
import { SwapInput } from '../themed/SwapInput'
import { ButtonBox } from '../themed/ThemedButtons'
import { AlertCardUi4 } from '../ui4/AlertCardUi4'
import { ButtonsViewUi4 } from '../ui4/ButtonsViewUi4'
import { ButtonUi4 } from '../ui4/ButtonUi4'

export interface SwapCreateParams {
  // The following props are used to populate the flip inputs
  fromWalletId?: string | undefined
  fromTokenId?: EdgeTokenId
  toWalletId?: string | undefined
  toTokenId?: EdgeTokenId

  // Display error message in an alert card
  errorDisplayInfo?: SwapErrorDisplayInfo
}

export interface SwapErrorDisplayInfo {
  message: string
  title: string
}

interface Props extends EdgeSceneProps<'swapCreate'> {}

interface State {
  nativeAmount: string
  fiatAmount: string
  nativeAmountFor: 'from' | 'to'
}

const defaultState: State = {
  nativeAmount: '0',
  fiatAmount: '0',
  nativeAmountFor: 'from'
}

export const SwapCreateScene = (props: Props) => {
  const { navigation, route } = props
  const { fromWalletId, fromTokenId = null, toWalletId, toTokenId = null, errorDisplayInfo } = route.params ?? {}
  const theme = useTheme()
  const dispatch = useDispatch()

  const [state, setState] = useState({
    ...defaultState
  })

  const fromInputRef = React.useRef<ExchangedFlipInputRef>(null)
  const toInputRef = React.useRef<ExchangedFlipInputRef>(null)

  const swapRequestOptions = useSwapRequestOptions()

  const account = useSelector(state => state.core.account)
  const currencyWallets = useWatch(account, 'currencyWallets')
  const exchangeInfo = useSelector(state => state.ui.exchangeInfo)

  const toWallet: EdgeCurrencyWallet | undefined = toWalletId == null ? undefined : currencyWallets[toWalletId]
  const fromWallet: EdgeCurrencyWallet | undefined = fromWalletId == null ? undefined : currencyWallets[fromWalletId]

  const toWalletName = toWallet == null ? '' : getWalletName(toWallet)
  const fromWalletName = fromWallet == null ? '' : getWalletName(fromWallet)
  const fromCurrencyCode = fromWallet == null ? '' : getCurrencyCode(fromWallet, fromTokenId)
  const toCurrencyCode = toWallet == null ? '' : getCurrencyCode(toWallet, toTokenId)

  const fromWalletSpecialCurrencyInfo = getSpecialCurrencyInfo(fromWallet?.currencyInfo.pluginId ?? '')
  const fromWalletBalanceMap = fromWallet?.balanceMap ?? new Map<string, string>()

  const fromHeaderText = fromWallet == null ? lstrings.select_src_wallet : fromWalletName
  const toHeaderText = toWallet == null ? lstrings.select_recv_wallet : toWalletName
  // Determines if a coin can have Exchange Max option
  const hasMaxSpend = fromWallet != null && fromWalletSpecialCurrencyInfo.noMaxSpend !== true

  const isNextHidden =
    // Don't show next button if the wallets haven't been selected:
    fromWallet == null ||
    toWallet == null ||
    // Don't show next button if the amount is zero:
    zeroString(state.nativeAmount) ||
    // Don't show next button if the amount exceeds the balance:
    checkAmountExceedsBalance()

  //
  // Callbacks
  //

  const checkDisableAsset = (disableAssets: DisableAsset[], walletId: string, tokenId: EdgeTokenId): boolean => {
    const wallet = currencyWallets[walletId] ?? { currencyInfo: {} }
    const walletPluginId = wallet.currencyInfo.pluginId
    const walletTokenId = tokenId
    for (const disableAsset of disableAssets) {
      const { pluginId, tokenId } = disableAsset
      if (pluginId !== walletPluginId) continue
      if (tokenId === walletTokenId) return true
      if (tokenId === 'allCoins') return true
      if (tokenId === 'allTokens' && walletTokenId != null) return true
    }
    return false
  }

  function checkAmountExceedsBalance(): boolean {
    // If no from wallet, return false:
    if (fromWallet == null) return false
    // We do not know what the from amount is if we are quoting "to" a
    // specific amount. Therefore we always return false in this case.
    if (state.nativeAmountFor === 'to') return false
    // Get the balance:
    const fromWalletBalance = fromWalletBalanceMap.get(fromTokenId) ?? '0'
    // If there is a balance and the amount is greater than the balance,
    // return true (which means amount exceeded balance).
    return gte(fromWalletBalance, '0') && gt(state.nativeAmount, fromWalletBalance)
  }

  const getQuote = (swapRequest: EdgeSwapRequest) => {
    if (exchangeInfo != null) {
      const disableSrc = checkDisableAsset(exchangeInfo.swap.disableAssets.source, swapRequest.fromWallet.id, fromTokenId)
      if (disableSrc) {
        showError(sprintf(lstrings.exchange_asset_unsupported, fromCurrencyCode))
        return
      }

      const disableDest = checkDisableAsset(exchangeInfo.swap.disableAssets.destination, swapRequest.toWallet.id, toTokenId)
      if (disableDest) {
        showError(sprintf(lstrings.exchange_asset_unsupported, toCurrencyCode))
        return
      }
    }
    // Clear the error state:
    navigation.setParams({
      ...route.params,
      errorDisplayInfo: undefined
    })

    // Start request for quote:
    navigation.navigate('swapProcessing', {
      swapRequest,
      swapRequestOptions,
      onCancel: () => {
        navigation.goBack()
      },
      onDone: quotes => {
        navigation.replace('swapConfirmation', {
          selectedQuote: quotes[0],
          quotes,
          onApprove: resetState
        })
      }
    })
  }

  const resetState = () => {
    setState(defaultState)
  }

  const showWalletListModal = async (whichWallet: 'from' | 'to') => {
    const result = await Airship.show<WalletListResult>(bridge => (
      <WalletListModal
        bridge={bridge}
        navigation={props.navigation}
        headerTitle={whichWallet === 'to' ? lstrings.select_recv_wallet : lstrings.select_src_wallet}
        showCreateWallet={whichWallet === 'to'}
        allowKeysOnlyMode={whichWallet === 'from'}
        filterActivation
      />
    ))
    if (result?.type === 'wallet') {
      const { walletId, tokenId } = result
      await handleSelectWallet(walletId, tokenId, whichWallet)
    }
  }

  //
  // Handlers
  //

  const handleFlipWalletPress = useHandler(() => {
    // Flip params:
    navigation.setParams({
      fromWalletId: toWalletId,
      fromTokenId: toTokenId,
      toWalletId: fromWalletId,
      toTokenId: fromTokenId,
      errorDisplayInfo
    })
    const newNativeAmountFor = state.nativeAmountFor === 'from' ? 'to' : 'from'
    // Clear amount input state:
    setState({
      ...state,
      nativeAmountFor: newNativeAmountFor
    })
    // Swap the amounts:
    const toAmount = newNativeAmountFor === 'to' ? state.fiatAmount : '0'
    const fromAmount = newNativeAmountFor === 'from' ? state.fiatAmount : '0'
    toInputRef.current?.setAmount('fiat', toAmount)
    fromInputRef.current?.setAmount('fiat', fromAmount)
  })

  const handleSelectWallet = useHandler(async (walletId: string, tokenId: EdgeTokenId, direction: 'from' | 'to') => {
    const params = {
      ...route.params,
      ...(direction === 'to'
        ? {
            toWalletId: walletId,
            toTokenId: tokenId
          }
        : {
            fromWalletId: walletId,
            fromTokenId: tokenId
          })
    }
    navigation.setParams(params)
    dispatch(updateMostRecentWalletsSelected(walletId, tokenId))
  })

  const handleMaxPress = useHandler(() => {
    if (toWallet == null) {
      showWarning(`${lstrings.exchange_select_receiving_wallet}`, { trackError: false })
      return
    }

    if (fromWallet == null) {
      // Shouldn't ever happen because max button UI is disabled when no
      // fromWallet is selected
      showWarning(`${lstrings.exchange_select_sending_wallet}`, { trackError: false })
      return
    }

    const request: EdgeSwapRequest = {
      fromTokenId: fromTokenId,
      fromWallet: fromWallet,
      nativeAmount: '0',
      quoteFor: 'max',
      toTokenId: toTokenId,
      toWallet: toWallet
    }

    getQuote(request)
  })

  const handleNext = useHandler(() => {
    // Should only happen if the user initiated the swap from the keyboard
    if (fromWallet == null || toWallet == null) return

    if (zeroString(state.nativeAmount)) {
      showToast(`${lstrings.no_exchange_amount}. ${lstrings.select_exchange_amount}.`)
      return
    }

    const request: EdgeSwapRequest = {
      fromTokenId: fromTokenId,
      fromWallet: fromWallet,
      nativeAmount: state.nativeAmount,
      quoteFor: state.nativeAmountFor,
      toTokenId: toTokenId,
      toWallet: toWallet
    }

    if (checkAmountExceedsBalance()) return

    getQuote(request)
  })

  const handleFromSelectWallet = useHandler(async () => {
    await showWalletListModal('from')
  })

  const handleToSelectWallet = useHandler(async () => {
    await showWalletListModal('to')
  })

  const handleFromAmountChange = useHandler((amounts: ExchangedFlipInputAmounts) => {
    setState({
      ...state,
      nativeAmount: amounts.nativeAmount,
      fiatAmount: amounts.fiatAmount,
      nativeAmountFor: 'from'
    })
    // Clear other input's amount:
    toInputRef.current?.setAmount('crypto', '0')
  })

  const handleToAmountChange = useHandler((amounts: ExchangedFlipInputAmounts) => {
    setState({
      ...state,
      nativeAmount: amounts.nativeAmount,
      fiatAmount: amounts.fiatAmount,
      nativeAmountFor: 'to'
    })
    // Clear other input's amount:
    fromInputRef.current?.setAmount('crypto', '0')
  })

  //
  // Render
  //

  const renderAlert = () => {
    const { minimumPopupModals } = fromWalletSpecialCurrencyInfo
    const primaryNativeBalance = fromWalletBalanceMap.get(fromTokenId) ?? '0'

    if (minimumPopupModals != null && primaryNativeBalance < minimumPopupModals.minimumNativeBalance) {
      return <AlertCardUi4 title={lstrings.request_minimum_notification_title} body={minimumPopupModals.alertMessage} type="warning" />
    }

    if (errorDisplayInfo != null) {
      return <AlertCardUi4 title={errorDisplayInfo.title} body={errorDisplayInfo.message} type="error" />
    }

    if (checkAmountExceedsBalance()) {
      return <AlertCardUi4 title={lstrings.exchange_insufficient_funds_title} body={lstrings.exchange_insufficient_funds_below_balance} type="error" />
    }

    return null
  }

  return (
    <SceneWrapper hasTabs hasNotifications scroll keyboardShouldPersistTaps="handled" padding={theme.rem(0.5)}>
      <EdgeAnim enter={fadeInUp60}>
        {fromWallet == null ? (
          <ButtonUi4 type="secondary" onPress={handleFromSelectWallet} marginRem={[1, 0]} label={lstrings.select_src_wallet} />
        ) : (
          <SwapInput
            ref={fromInputRef}
            heading={lstrings.exchange_title_sending}
            forceField="fiat"
            walletPlaceholderText={fromHeaderText}
            keyboardVisible={false}
            onAmountChanged={handleFromAmountChange}
            onNext={handleNext}
            onSelectWallet={handleFromSelectWallet}
            placeholders={[isNextHidden ? lstrings.string_tap_to_edit : fromCurrencyCode, isNextHidden ? '' : lstrings.string_tap_next_for_quote]}
            tokenId={fromTokenId}
            wallet={fromWallet}
          />
        )}
      </EdgeAnim>
      <EdgeAnim>
        <LineTextDivider lowerCased>
          <ButtonBox onPress={handleFlipWalletPress} paddingRem={[0, 0.5]}>
            <SwapVerticalIcon color={theme.iconTappable} size={theme.rem(2)} />
          </ButtonBox>
          {hasMaxSpend ? (
            <MaxButtonContainerView>
              <EdgeTouchableOpacity onPress={handleMaxPress}>
                <MaxButtonText>{lstrings.string_max_cap}</MaxButtonText>
              </EdgeTouchableOpacity>
            </MaxButtonContainerView>
          ) : null}
        </LineTextDivider>
      </EdgeAnim>
      <EdgeAnim enter={fadeInDown30}>
        {toWallet == null ? (
          <ButtonUi4 type="secondary" onPress={handleToSelectWallet} marginRem={[1, 0]} label={lstrings.select_recv_wallet} />
        ) : (
          <SwapInput
            ref={toInputRef}
            forceField="fiat"
            walletPlaceholderText={toHeaderText}
            keyboardVisible={false}
            onAmountChanged={handleToAmountChange}
            onNext={handleNext}
            onSelectWallet={handleToSelectWallet}
            placeholders={[isNextHidden ? lstrings.string_tap_to_edit : toCurrencyCode, isNextHidden ? '' : lstrings.string_tap_next_for_quote]}
            tokenId={toTokenId}
            wallet={toWallet}
            heading={lstrings.exchange_title_receiving}
          />
        )}
      </EdgeAnim>
      <EdgeAnim enter={fadeInDown60}>{renderAlert()}</EdgeAnim>
      <EdgeAnim enter={fadeInDown90}>
        {isNextHidden ? null : <ButtonsViewUi4 primary={{ label: lstrings.string_next_capitalized, onPress: handleNext }} parentType="scene" />}
      </EdgeAnim>
    </SceneWrapper>
  )
}

const MaxButtonContainerView = styled(View)(theme => ({
  position: 'absolute',
  right: theme.rem(1),
  top: -theme.rem(0.5)
}))

const MaxButtonText = styled(Text)(theme => ({
  color: theme.escapeButtonText,
  fontFamily: theme.fontFaceDefault,
  fontSize: theme.rem(0.75),
  includeFontPadding: false
}))
