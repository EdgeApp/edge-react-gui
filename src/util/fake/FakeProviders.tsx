import * as React from 'react'
import { Metrics, SafeAreaProvider } from 'react-native-safe-area-context'
import { Provider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'

import { rootReducer, RootState } from '../../reducers/RootReducer'

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>
    }
  : T

export type FakeState = DeepPartial<RootState>

interface Props {
  children: React.ReactNode
  initialState?: FakeState
}

export function FakeProviders(props: Props) {
  const { children, initialState = {} } = props

  const store = React.useMemo(() => createStore(rootReducer, initialState as any, applyMiddleware(thunk)), [initialState])
  return (
    <SafeAreaProvider initialMetrics={initialMetrics}>
      <Provider store={store}>{children}</Provider>
    </SafeAreaProvider>
  )
}

// These match what our old react-native-safe-area-view library returned:
const initialMetrics: Metrics = {
  frame: { height: 1334, width: 750, x: 0, y: 0 },
  insets: { bottom: 0, left: 0, right: 0, top: 20 }
}
