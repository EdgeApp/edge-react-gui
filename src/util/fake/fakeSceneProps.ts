import { AppParamList, EdgeSceneProps, NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener() {
    return () => undefined
  },
  removeListener() {},
  isFocused() {
    return true
  },

  navigate() {},
  push() {},
  replace() {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  dispatch() {},
  reset() {},
  canGoBack() {
    return true
  },
  getId() {
    return undefined
  },
  getState() {
    return {
      key: 'foo',
      index: 0,
      routes: [],
      routeNames: [],
      type: 'bar',
      stale: false
    }
  },
  getParent() {
    throw new Error('not implemented')
  },
  setOptions() {}
}

export function fakeSceneProps<Name extends keyof AppParamList>(name: Name, params: AppParamList[Name]): EdgeSceneProps<Name> {
  return {
    navigation: fakeNavigation,
    route: {
      name: name,
      key: name + '-0',
      params
    } as any
  }
}
