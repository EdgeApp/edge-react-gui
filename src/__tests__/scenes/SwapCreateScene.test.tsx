import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import TestRenderer from 'react-test-renderer'

import { SwapCreateScene } from '../../components/scenes/SwapCreateScene'
import { FakeProviders, FakeState } from '../../util/fake/FakeProviders'
import { fakeRootState } from '../../util/fake/fakeRootState'
import { fakeSceneProps } from '../../util/fake/fakeSceneProps'

describe('SwapCreateScene', () => {
  it('should render with loading props', () => {
    const rootState: FakeState = { ...fakeRootState }

    const renderer = TestRenderer.create(
      <FakeProviders initialState={rootState}>
        <SwapCreateScene {...fakeSceneProps('swapCreate', {})} />
      </FakeProviders>
    )

    expect(renderer.toJSON()).toMatchSnapshot()
    renderer.unmount()
  })
})
