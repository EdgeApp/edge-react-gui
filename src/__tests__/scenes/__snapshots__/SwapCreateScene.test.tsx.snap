// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapCreateScene should render with loading props 1`] = `
[
  <BVLinearGradient
    colors={
      [
        4279839052,
        4280174446,
      ]
    }
    endPoint={
      {
        "x": 0,
        "y": 0,
      }
    }
    locations={null}
    onLayout={[Function]}
    startPoint={
      {
        "x": 1,
        "y": 0,
      }
    }
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={0}
      bbWidth={0}
      focusable={false}
      height={0}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        }
      }
      vbHeight={0}
      vbWidth={0}
      width={0}
    >
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGDefs />
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          opacity={0.1}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </BVLinearGradient>,
  <RCTScrollView
    collapsable={false}
    contentContainerStyle={
      {
        "paddingBottom": 74.85492957746479,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 64,
      }
    }
    keyboardShouldPersistTaps="handled"
    onScroll={[Function]}
    scrollEventThrottle={0.0001}
    scrollIndicatorInsets={
      {
        "right": 1,
      }
    }
    style={
      {
        "height": 1334,
        "maxHeight": 1334,
        "padding": 11,
        "width": 750,
      }
    }
  >
    <View>
      <View
        entering={BaseAnimationMock {}}
        layout={BaseAnimationMock {}}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "alignSelf": "center",
              "flexBasis": "auto",
              "flexGrow": 0,
              "flexShrink": 0,
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 22,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 22,
            }
          }
        >
          <BVLinearGradient
            colors={
              [
                872415231,
                872415231,
              ]
            }
            endPoint={
              {
                "x": 1,
                "y": 1,
              }
            }
            locations={null}
            startPoint={
              {
                "x": 0,
                "y": 0,
              }
            }
            style={
              [
                {
                  "alignItems": "center",
                  "borderRadius": 34,
                  "flexDirection": "row",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "justifyContent": "center",
                },
                {
                  "alignSelf": "center",
                },
                {
                  "height": 67,
                  "paddingHorizontal": 34,
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              minimumFontScale={0.65}
              numberOfLines={1}
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "NunitoSans-Regular",
                    "fontSize": 22,
                    "includeFontPadding": false,
                  },
                  [
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "NunitoSans-Regular",
                      "fontSize": 22,
                    },
                    null,
                  ],
                  null,
                ]
              }
            >
              Select Source Wallet
            </Text>
          </BVLinearGradient>
        </View>
      </View>
      <View
        layout={BaseAnimationMock {}}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "flexGrow": 1,
              "justifyContent": "space-between",
              "marginVertical": 11,
              "paddingHorizontal": 11,
            }
          }
        >
          <View
            style={
              {
                "borderBottomColor": "#e27226",
                "borderBottomWidth": 1,
                "flex": 1,
              }
            }
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={false}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "marginBottom": 0,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 0,
                "opacity": 1,
                "paddingBottom": 0,
                "paddingLeft": 11,
                "paddingRight": 11,
                "paddingTop": 0,
              }
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": undefined,
                    "fontSize": 12,
                  },
                  {
                    "color": "#e27226",
                    "fontSize": 45,
                  },
                  {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              îº´
            </Text>
          </View>
          <View
            style={
              {
                "borderBottomColor": "#e27226",
                "borderBottomWidth": 1,
                "flex": 1,
              }
            }
          />
        </View>
      </View>
      <View
        entering={BaseAnimationMock {}}
        layout={BaseAnimationMock {}}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "alignSelf": "center",
              "flexBasis": "auto",
              "flexGrow": 0,
              "flexShrink": 0,
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 22,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 22,
            }
          }
        >
          <BVLinearGradient
            colors={
              [
                872415231,
                872415231,
              ]
            }
            endPoint={
              {
                "x": 1,
                "y": 1,
              }
            }
            locations={null}
            startPoint={
              {
                "x": 0,
                "y": 0,
              }
            }
            style={
              [
                {
                  "alignItems": "center",
                  "borderRadius": 34,
                  "flexDirection": "row",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "justifyContent": "center",
                },
                {
                  "alignSelf": "center",
                },
                {
                  "height": 67,
                  "paddingHorizontal": 34,
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              minimumFontScale={0.65}
              numberOfLines={1}
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "NunitoSans-Regular",
                    "fontSize": 22,
                    "includeFontPadding": false,
                  },
                  [
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "NunitoSans-Regular",
                      "fontSize": 22,
                    },
                    null,
                  ],
                  null,
                ]
              }
            >
              Select Receiving Wallet
            </Text>
          </BVLinearGradient>
        </View>
      </View>
      <View
        entering={BaseAnimationMock {}}
        layout={BaseAnimationMock {}}
      />
      <View
        entering={BaseAnimationMock {}}
        layout={BaseAnimationMock {}}
      />
    </View>
  </RCTScrollView>,
  <View
    footerHeight={0}
    footerOpenRatio={
      {
        "value": 1,
      }
    }
    hasTabs={true}
    insetBottom={0}
    onLayout={[Function]}
    style={
      [
        {
          "alignSelf": "center",
          "bottom": 0,
          "justifyContent": "flex-end",
          "paddingHorizontal": 11,
          "position": "absolute",
        },
        {
          "transform": [
            {
              "translateY": NaN,
            },
          ],
        },
      ]
    }
  >
    <View
      entering={BaseAnimationMock {}}
      exiting={BaseAnimationMock {}}
      layout={BaseAnimationMock {}}
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#1c385d",
              "borderRadius": 11,
              "flexDirection": "row",
              "justifyContent": "center",
              "margin": 11,
              "marginTop": 0,
              "padding": 11,
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0.7,
              "shadowRadius": 11,
            },
            {
              "opacity": 1,
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "flexShrink": 1,
              "justifyContent": "center",
              "margin": 6,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "height": 56,
                  "marginRight": 6,
                  "width": 56,
                },
              ]
            }
          >
            <FastImageView
              resizeMode="cover"
              source={
                {
                  "uri": "https://content.edge.app/notifications/icon-lock.png",
                }
              }
              style={
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
          </View>
          <View
            style={
              {
                "flexShrink": 1,
                "width": "100%",
              }
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              minimumFontScale={0.65}
              numberOfLines={1}
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "NunitoSans-Regular",
                    "fontSize": 22,
                    "includeFontPadding": false,
                  },
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "NunitoSans-Bold",
                    "fontSize": 17,
                    "marginHorizontal": 6,
                  },
                  null,
                ]
              }
              type="info"
            >
              IP Validation Protection
            </Text>
            <Text
              adjustsFontSizeToFit={true}
              minimumFontScale={0.65}
              numberOfLines={3}
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "NunitoSans-Regular",
                    "fontSize": 22,
                    "includeFontPadding": false,
                  },
                  {
                    "color": "#A5D4FF",
                    "fontSize": 17,
                    "marginHorizontal": 6,
                  },
                  null,
                ]
              }
              type="info"
            >
              Coinhub Bitcoin Wallet prevents logins from unrecognized IP addresses. Tap to learn more.
            </Text>
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "margin": 6,
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#e27226",
                  "fontSize": 28,
                },
                undefined,
                {
                  "fontFamily": "anticon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            î³
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
